packages/service/src/utils.ts

```diff
@@ -263,3 +278,274 @@ export function determineNativeContext(
+    });
+    `
+    return it
+}
+
+/**
+ * Write the test file
+ */
+function writeTestFile(directoryPath: string, fileID: string, log: Logger, testContent: string) {
+    const filePath = join(directoryPath, `${fileID}.test.js`)
+    try {
+        writeFileSync(filePath, testContent)
+        log.info(`Test file created at: ${filePath}`)
+    } catch (err) {
+        console.error(`It seems that the writing the file to '${filePath}' didn't succeed due to the following error: ${err}`)
+        process.exit(1)
+    }
+}
+
+/**
+ * Create the test content
+ */
+function createTestContent ({ clip, clipSelector, stories, storybookUrl }:CreateTestContent):string {
+    return stories.reduce((acc, storyData) => acc + itFunction(clip, clipSelector, storyData, storybookUrl), '')
+}
+
+/**
+ * Create the file data
+ */
+function createFileData (describeTitle: string, testContent: string):string {
+    return `\ndescribe(\`${describeTitle}\`, () => {\n    ${testContent}\n});\n`
+}
+
+/**
+ * Create the test files
+ */
+export function createTestFiles({
+    clip,
+    clipSelector,
+    directoryPath,
+    log,
+    numShards,
+    storiesJson,
+    storybookUrl,
+}: CreateTestFileOptions) {
+    const storiesArray = Object.values(storiesJson)
+        // By default only keep the stories, not the docs
+        .filter((storyData: StorybookData) => storyData?.type === 'story' || !storyData.parameters?.docsOnly)
+    const fileNamePrefix = 'visual-storybook'
+
+    if (numShards === 1){
+        const testContent = createTestContent({ clip, clipSelector, stories: storiesArray, storybookUrl })
+        const fileData = createFileData('All stories', testContent)
+        writeTestFile(directoryPath, `${fileNamePrefix}-1-1`, log, fileData)
+    } else {
+        const totalStories = storiesArray.length
+        const storiesPerShard = Math.ceil(totalStories / numShards)
+
+        for (let shard = 0; shard < numShards; shard++) {
+            const startIndex = shard * storiesPerShard
+            const endIndex = Math.min(startIndex + storiesPerShard, totalStories)
+            const shardStories = storiesArray.slice(startIndex, endIndex)
+            const testContent = createTestContent({ clip, clipSelector, stories: shardStories, storybookUrl })
+            const fileId = `${fileNamePrefix}-${shard + 1}-${numShards}`
+            const describeTitle = `Shard ${shard + 1} of ${numShards}`
+            const fileData = createFileData(describeTitle, testContent)
+
+            writeTestFile(directoryPath, fileId, log, fileData)
+        }
+    }
+}
+
+/**
+ * Create the storybook capabilities based on the specified browsers
+ */
+export function createStorybookCapabilities(capabilities: Capabilities.RemoteCapabilities, log: Logger) {
+    const isHeadless = process.argv.includes('--headless')
+    const browsers = process.argv.includes('--browsers') ? process.argv[process.argv.indexOf('--browsers') + 1].split(',') : ['chrome']
```

Comment:
Do we want to decouple the capabilities we use to make the screenshot tests with the one defined in the `wdio.conf.ts`? I see advantages and disadvantages in both:

- disadvantage: so far every WebdriverIO use case is connected to the capabilities defined in the `wdio.conf.ts` file, decoupling from that could get confusing to folks especially if the config is composed with special logic and there are multiple places where capabilities are defined for different purposes
- advantages: it is indeed nice to combine storybook tests with other types of test and have a single wdio command that is able to run them all

Right now it is not unusual to have different WebdriverIO configs per usecase, e.g. we have e2e and component tests that require a different config. I don't see a reason why folks wouldn't have a:

- `wdio.e2e.conf.ts` for e2e tests
- `wdio.unit.conf.ts` for unit/component tests in the browser
- `wdio.visual.conf.ts` for storybook tests

What do you think?

Reply:
we could keep the current caps and only overwrite it when it's provided through the command line. There might be some challenges with real Mobile browsers due to the localhost urls